User Authentication and Authorization

sea-gull-diana opened the pull request:
## What does this PR do ?
Implements endpoints to register and authenticate user in `UsersModule` and `AuthModule` respectively. Adds authorization which depends on the user table's `isAdmin` attribute.

### How should this be manually tested?

  - Step 1 : launch the app using docker-compose or with `npm start` (see README).
  - Step 2 : send a POST request to `localhost:3000/users` with the following JSON body:
  ```
  {
      "username": "test",
      "email": "test@test.com",
      "password": "Test@12345"
  }
  ```
  A non-admin user will be created (you can change the property `isAdmin` to true in the `create` method of `UsersService` if you want to create an admin user).
  - Step 3 : send the same POST request, but to `localhost:3000/auth` to get an access token (valid for half an hour).
  - Step 4 : use the token as Bearer token in Authorization header to send requests to protected routes (all routes which are not marked with `Public()` decorator.

### TODO
Adding email verification and password reset features with the use of `MailModule`.

### Other changes
The app has custom decorators `@Public()` and `@AdminOnly()` that you can add to controllers to make routes accessible to unauthenticated users or accessible to admin only.

sea-gull-diana committed:
feat: add database to authentication module

sea-gull-diana committed:
feat: validate email, password and username forms on user create

mfernd commented:
Thanks for your PR!

So I have a few questions for this:

1. Could we have a refresh token route for refreshing our access token without re-login? Like `/auth/refresh`
2. If there is no refresh token, can you extend the duration of the access token to a more reasonable time, like `6h`?
3. Is it not better to move the `POST` on `/users` for creating an account, to the auth module, with a route `/auth/register`?  
(same for login, not on `/auth` but `/auth/login`)
4. Is it not better to use cookies instead of Bearer token?

(Note that a refresh token should live for something like a week.)

Some recommendations on refactoring:

1. For the env variables there is a module in nestjs called [ConfigModule](https://docs.nestjs.com/techniques/configuration)
2. For errors, nestjs has [HttpException](https://docs.nestjs.com/exception-filters#throwing-standard-exceptions), so I don't know if re-implementing with `ResponseError` and `ResponseSuccess` is really useful

sea-gull-diana committed:
fix: move user creation route to AuthModule, remove IResponse

BioTheWolff commented on .env.example:
```diff
@@ -4,6 +4,9 @@
 APP_PORT=80
 
+# A secret string used to encrypt data into JWT tokens on sign in
+JWT_SECRET="A secret string for the JWT token";
```
```suggestion
JWT_SECRET="A secret string for the JWT token"
```

BioTheWolff commented on src/auth/auth.controller.ts:
```diff
@@ -0,0 +1,44 @@
+    @Public()
+    @Post('login')
+    async login(@Body() loginUserDto: LoginUserDto): Promise<AccessToken> {
+        return this.authService.signIn(loginUserDto);
```
I recommend renaming the service's method to have the same name as the controller route name

BioTheWolff commented on src/auth/auth.controller.ts:
```diff
@@ -0,0 +1,44 @@
+    // This action adds a new user
+    @Public()
+    @Post('register')
+    async create(@Body() createUserDto: CreateUserDto): Promise<SanitizedUser> {
```
```suggestion
    async register(@Body() createUserDto: CreateUserDto): Promise<SanitizedUser> {
```

BioTheWolff commented on src/auth/auth.controller.ts:
```diff
@@ -0,0 +1,44 @@
+    @Public()
+    @Post('register')
+    async create(@Body() createUserDto: CreateUserDto): Promise<SanitizedUser> {
+        if (!await this.usersService.validateEmail(createUserDto.email)) {
+            throw new HttpException("Registration error - email is invalid or already used", HttpStatus.BAD_REQUEST);
+        }
+
+        if (!await this.usersService.validateUsername(createUserDto.username)) {
+            throw new HttpException("Registration error - username is invalid or already used", HttpStatus.BAD_REQUEST);
+        }
+
+        if (!await this.usersService.validatePassword(createUserDto.password)) {
+            throw new HttpException("Registration error - password is invalid", HttpStatus.BAD_REQUEST);
+        }
+
+        return await this.usersService.create(createUserDto);
```
Other service validation should always be made in the service, not in the controller. I recommend adding a method named `register` in `AuthService` and handling it there.

sea-gull-diana replied:
Ok, I'll move all usage of userService to authService :)

BioTheWolff commented on src/auth/auth.module.ts:
```diff
@@ -0,0 +1,22 @@
+  providers: [AuthService, PrismaService],
+  imports: [
+    UsersModule,
+    ConfigModule.forRoot(),
```
`ConfigModule` is already imported in root `AppModule`, it should be removed from here.

```suggestion

```

BioTheWolff commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,40 @@
+  private async validateEmail(email: string) {
+    // TODO: check that such email address doesn't already exist in the db,
+    // send a validation email to this address
+  }
```
Suggesting to remove unused method

```suggestion
```

BioTheWolff commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,40 @@
+
+  // Sign user in by email and password
+  async signIn(credentials: LoginUserDto): Promise<AccessToken> {
+    const user = await this.prisma.user.findUnique({
+      where: {email: credentials.email}
+    });
+    await this.usersService.findOne({email: credentials.email});
```
`AuthService` does not have responsibility to use `prisma.user`, as it *should* be reserved to `UsersService`. I recommend using a findOne-like.

Note: as `findOne` is returning a `SanitizedUser`, you could use a `findOneUnsanitized` that would return the full user, which would include the password.

```suggestion
    const user = await this.usersService.findOne({email: credentials.email});
```

sea-gull-diana replied:
Ok, I'll add an `findOneUnsanitized` method into userService then.

BioTheWolff commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,40 @@
+    });
+    await this.usersService.findOne({email: credentials.email});
+    const passwordHash = user['password'];
+    const isCorrectPassword = await bcrypt.compare(credentials.password, passwordHash || "");
```
Why use an empty password?

```suggestion
    const isCorrectPassword = await bcrypt.compare(credentials.password, passwordHash);
```

BioTheWolff commented on src/auth/guards/roles.guard.ts:
```diff
@@ -0,0 +1,21 @@
+    const { user } = context.switchToHttp().getRequest();
+    return user.isAdmin;
+  }
+}
```
Add trailing newline to file

```suggestion
}

```

BioTheWolff commented on src/users/users.service.ts:
```diff
@@ -1,35 +1,69 @@
-        password: this.passwd_encrypt(user.password),
-        isAdmin: false,
+        password: await this.passwd_encrypt(user.password),
+        isAdmin: true,
```
Users should definitely not be admins by default. I think this is something that's been forgotten during a commit.

sea-gull-diana replied:
Yep, well noticed, I forgot to change it back, thanks.

BioTheWolff commented on src/users/users.service.ts:
```diff
@@ -1,35 +1,69 @@
     }));
   }
 
+  async validateEmail(email: string): Promise<boolean> {
+    // Verify if email form is valid
+    const expression: RegExp = EMAIL_REGEX;
+    const isValidRegex: boolean = expression.test(email);
+
+    if (isValidRegex) {
+        // Check that email address doesn't already exist in the db
+        const user = await this.findOne({email});
+        return !user;
+    }
+    return false;
+  }
+
+  async validateUsername(username: string): Promise<boolean> {
+    // Verify if username form is valid
+    const expression: RegExp = USERNAME_REGEX;
+    const isValidRegex: boolean = expression.test(username);
+
+    if (isValidRegex) {
+        // Check that username doesn't already exist in the db
+        const user = await this.findOne({username});
+        return !user;
+    }
+    return false;
+  }
+
+  async validatePassword(password: string): Promise<boolean> {
+    const expression: RegExp = PASSWORD_REGEX;
+    return(expression.test(password));
+  }
```
This could be handled in the DTOs using NestJS's `class-validator` and the `whitelist` property in the `main.ts` file. See [Auto-validation -  NestJS Documentation](https://docs.nestjs.com/techniques/validation#auto-validation)

sea-gull-diana replied:
Noted. I'll add a validation pipeline.

BioTheWolff commented on src/users/users.service.ts:
```diff
@@ -1,35 +1,69 @@
   }
 
+  private async passwd_encrypt(password: string): Promise<string> {
+    const saltOrRounds = await bcrypt.genSalt(10);
+    const hash = await bcrypt.hash(password, saltOrRounds);
+    return hash;
```
```suggestion
    return await bcrypt.hash(password, 10);
```

sea-gull-diana commented:
> Thanks for your PR!
> 
> So I have a few questions for this:
> 
> 1. Could we have a refresh token route for refreshing our access token without re-login? Like `/auth/refresh`
> 2. If there is no refresh token, can you extend the duration of the access token to a more reasonable time, like `6h`?
> 3. Is it not better to move the `POST` on `/users` for creating an account, to the auth module, with a route `/auth/register`?
>    (same for login, not on `/auth` but `/auth/login`)
> 4. Is it not better to use cookies instead of Bearer token?
> 
> (Note that a refresh token should live for something like a week.)
> 
> Some recommendations on refactoring:
> 
> 1. For the env variables there is a module in nestjs called [ConfigModule](https://docs.nestjs.com/techniques/configuration)
> 2. For errors, nestjs has [HttpException](https://docs.nestjs.com/exception-filters#throwing-standard-exceptions), so I don't know if re-implementing with `ResponseError` and `ResponseSuccess` is really useful

**Thanks for your feedback!**

- Refresh tokens are work in progress, I have changed token expiration time to 6 hours for now.
- The routes I changed as you asked, we now have `/auth/register` and `/auth/login`.
- Good question about cookies, they are indeed considered by some as more secure. I'll make sure to use them for the refresh token.
- Regarding the ConfigModule, I am already using it, so feel free to inject `ConfigService` when you need to use environment variables.
- I have also removed `ResponseError` and `ResponseSuccess` types, they are indeed not very important.

BioTheWolff commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,40 @@
+
+    return {accessToken: await this.jwtService.signAsync(payload)};
+  }
+}
```
Add newline to file

```suggestion
}

```

BioTheWolff reviewed (changes requested):
Overall a good PR, with only a few things to change!

The next step, later in the future, would be to use Passport for authentication - for now, using NestJS only is good.

I want to stress that changing `isAdmin` back to false is really important, please take care of that ASAP.

BioTheWolff commented on package-lock.json:
Should we not keep the package lock? It is needed for CI and clean builds.

sea-gull-diana replied:
Not sure what you mean. I do indeed think that package-lock.json shouldn't be removed from commits. Did I accidentally exclude it somewhere?

BioTheWolff replied:
I misunderstood the heavy refactor for a complete deletion, my bad

sea-gull-diana committed:
fix: apply suggestions from code review

Co-authored-by: BioTheWolff <biothewolff@gmx.fr>

sea-gull-diana committed:
fix: set isAdmin to false by default, move validation on register to userService, add findOneUnsanitized method

sea-gull-diana committed:
feat: add NestJS validation pipeline with class-validator on user creation

sea-gull-diana committed:
feat: add class-validator decorators to all dto classes

BioTheWolff reviewed (approved):
LGTM, well done!
